safehttp/server.go

```diff
@@ -0,0 +1,189 @@
+	"crypto/tls"
+	"errors"
+	"net"
+	"net/http"
+	"time"
+)
+
+// Server is a safe wrapper for a standard HTTP server.
+// The zero value is safe and ready to use and will apply safe defaults on serving.
+// Changing any of the fields after the server has been started is a no-op.
+type Server struct {
+	// Addr optionally specifies the TCP address for the server to listen on,
+	// in the form "host:port". If empty, ":http" (port 80) is used.
+	// The service names are defined in RFC 6335 and assigned by IANA.
+	// See net.Dial for details of the address format.
+	Addr string
+
+	// Mux is the ServeMux to use for the current server. A nil Mux is invalid.
+	Mux *ServeMux
+
+	// TODO(empijei): potentially consider exposing ReadHeaderTimeout for
+	// fine-grained handling (e.g. websocket endpoints).
+
+	// ReadTimeout is the maximum duration for reading the entire
+	// request, including the body.
+	ReadTimeout time.Duration
+
+	// WriteTimeout is the maximum duration before timing out
+	// writes of the response. It is reset whenever a new
+	// request's header is read.
+	WriteTimeout time.Duration
+
+	// IdleTimeout is the maximum amount of time to wait for the
+	// next request when keep-alives are enabled.
+	IdleTimeout time.Duration
+
+	// MaxHeaderBytes controls the maximum number of bytes the
+	// server will read parsing the request header's keys and
+	// values, including the request line. It does not limit the
+	// size of the request body.
+	MaxHeaderBytes int
+
+	// TLSConfig optionally provides a TLS configuration for use
+	// by ServeTLS and ListenAndServeTLS. Note that this value is
+	// cloned on serving, so it's not possible to modify the
+	// configuration with methods like tls.Config.SetSessionTicketKeys.
+	TLSConfig *tls.Config
+
+	// OnShutdown is a slice of functions to call on Shutdown.
+	// This can be used to gracefully shutdown connections that have undergone
+	// ALPN protocol upgrade or that have been hijacked.
+	// These functions should start protocol-specific graceful shutdown, but
+	// should not wait for shutdown to complete.
+	OnShudown []func()
+
+	// DisableKeepAlives controls whether HTTP keep-alives should be disabled.
+	DisableKeepAlives bool
+
+	srv     *http.Server
+	started bool
+}
+
+func (s *Server) buildStd() error {
+	if s.started {
+		return errors.New("server already started")
+	}
+	if s.srv != nil {
+		// Server was already built
+		return nil
+	}
+	if s.Mux == nil {
+		return errors.New("building server without a mux")
+	}
+
+	srv := &http.Server{
+		Addr:           s.Addr,
+		Handler:        s.Mux,
+		ReadTimeout:    5 * time.Second,
+		WriteTimeout:   5 * time.Second,
+		IdleTimeout:    120 * time.Second,
+		MaxHeaderBytes: 10 * 1024,
+	}
+	if s.ReadTimeout != 0 {
+		srv.ReadTimeout = s.ReadTimeout
+	}
+	if s.WriteTimeout != 0 {
+		srv.WriteTimeout = s.WriteTimeout
+	}
+	if s.IdleTimeout != 0 {
+		srv.IdleTimeout = s.IdleTimeout
+	}
+	if s.MaxHeaderBytes != 0 {
+		srv.MaxHeaderBytes = 10 * 1024
+	}
```

Comment:
Why is this set this way?

Reply:
The default is a very very high and unreasonable value. 10K should be enough as a default. According to [this](http://dev.chromium.org/spdy/spdy-whitepaper) the average is about 2K, so I decided to leave a bit more room for it.

Reply:
But, if I set `s.MaxHeaderBytes` to `123`, why does it get overridden to `10*1024`?